src/design/interface.rs

```diff
@@ -0,0 +1,136 @@
+//! This module contains structures that represent Tydi interfaces on Streamlets,
+//! or other abstractions.
+
+use crate::design::{InterfaceKey, TypeRef};
+use crate::{Document, Error, Identify, Result, Reverse, Reversed};
+use std::convert::TryInto;
+use std::fmt::{Display, Formatter};
+use std::str::FromStr;
+
+/// Streamlet interface mode.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum Mode {
+    /// The interface is an output of the streamlet.
+    Out,
+    /// The interface is an input of the streamlet.
+    In,
+}
+
+impl Reversed for Mode {
+    fn reversed(&self) -> Self {
+        match self {
+            Mode::Out => Mode::In,
+            Mode::In => Mode::Out,
+        }
+    }
+}
+
+impl Display for Mode {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::result::Result<(), std::fmt::Error> {
+        write!(
+            f,
+            "{}",
+            match self {
+                Mode::Out => "Out",
+                Mode::In => "In",
+            }
+        )
+    }
+}
+
+impl FromStr for Mode {
+    type Err = Error;
+
+    fn from_str(input: &str) -> Result<Self> {
+        match input {
+            "in" => Ok(Mode::In),
+            "out" => Ok(Mode::Out),
+            _ => Err(Error::InvalidArgument(format!(
+                "{} is not a valid interface Mode. Expected \"in\" or \"out\"",
+                input
+            ))),
+        }
+    }
+}
+
+/// A Streamlet interface.
+///
+/// The names "clk" and "rst" are reserved.
+#[derive(Clone, Debug, PartialEq)]
```

Comment:
This `PartialEq` can be confusing. The implementation should be moved to the context where a reference to the actual project exists, to prevent false positives (if keys match for different libraries).

Reply:
The type reference includes the library within the project context, and there is no larger context intended for now.